### ISSUE TITLE
Problems with JPA and "bytea" columns

### ISSUE THREAD

**Issue by rdifalco** [impossibl/pgjdbc-ng#71]
Under JPA 2.0 one should be able to do this:

```
@Entity
public class Credential {

    @Id
    private Long id;

    @Column( length = 8 )
    private byte[] salt;

    @Column( length = 32 )
    private byte[] hash;
```

If you use this driver you will get an exception like this:

Caused by: java.lang.ClassCastException: [B cannot be cast to java.io.InputStream
    at com.impossibl.postgres.system.procs.Bytes$BinEncoder.length(Bytes.java:169)
    at com.impossibl.postgres.protocol.v30.ProtocolImpl.lengthOfParams(ProtocolImpl.java:723)
    at com.impossibl.postgres.protocol.v30.ProtocolImpl.writeBind(ProtocolImpl.java:507)
    at com.impossibl.postgres.protocol.v30.BindExecCommandImpl.execute(BindExecCommandImpl.java:319)
    at com.impossibl.postgres.protocol.v30.ProtocolImpl.execute(ProtocolImpl.java:369)
    at com.impossibl.postgres.jdbc.PGConnectionImpl.execute(PGConnectionImpl.java:428)
    at com.impossibl.postgres.jdbc.PGPreparedStatement.executeBatch(PGPreparedStatement.java:328)

I am using Spring + Hibernate for my JPA implementation. I tested this against 0.30 (the latest in maven at the time of this writing).

**Comment by depsypher** [comment]
I'm in the same boat. Would love to switch to pgjdbc-ng once this is resolved!

**Comment by davecramer** [comment]
Does the postgresql driver support this ?

Dave Cramer

On Mon, Mar 17, 2014 at 11:35 AM, Ray Vanderborght <notifications@github.com

**Comment by bvarner** [comment]
Yes, it does. I've used that extensively with JPA2 and Hibernate.

**Comment by kdubb** [comment]
FYI, I've been on vacation for the last 10 days.  Will get to this today as it appears to be an easy fix.

**Comment by kdubb** [comment]
Can you provide me a complete sample maven project I can use to test this?  It would be very helpful

**Comment by rdifalco** [comment]
I'm under deadlines right now so it might take me a few days. But really
you can just take ANY JPA/Hibernate sample project and add an Entity that
has a byte[] field that maps to a bytea column in Postgres. Then persist
it.

**Comment by kdubb** [comment]
@rdifalco  I did that... it seems to work just fine.  Hibernate 4.3.4 & pgjdbc-ng 0.3

**Comment by rdifalco** [comment]
Hmmm...and with JPA annotations (not Hibernate). So for example just like
this?

@Entity
public class Credential {

```
@Id
private Long id;

@Column( length = 8 )
private byte[] salt;

@Column( length = 32 )
private byte[] hash;
```

...

I can try it again but there is nothing fancy in my implementation and for
me (using hibernate 4x JPA entity manager) this results in the driver
trying to cast the byte[] column to an InputStream.

**Comment by kdubb** [comment]
```
package hibernate;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;

@Entity
public class Event {

  private long id;
  private byte[] data;

  @Id
  @GeneratedValue
  public long getId() {
    return id;
  }

  public void setId(long id) {
    this.id = id;
  }

  @Column(length = 32)
  public byte[] getData() {
    return data;
  }

  public void setData(byte[] data) {
    this.data = data;
  }

}
```

**Comment by kdubb** [comment]
```
    Event event = new Event();
    event.setData(new byte[] {1, 2, 3, 4, 5});

    EntityManager entityManager = entityManagerFactory.createEntityManager();
    entityManager.getTransaction().begin();
    entityManager.persist(event);
    entityManager.getTransaction().commit();

    entityManager.clear();

    Event event2 = entityManager.find(Event.class, event.getId());
```

**Comment by rdifalco** [comment]
I figured it out! Enable batch statements by adding this to your
persistence.xml:

```
        <property name="hibernate.jdbc.batch_size" value="50"/>
```

**Comment by kdubb** [comment]
Fixed in ba2def20943cc2a5cb2d720077476ccb7d5bb62b

**Comment by kdubb** [comment]
Let me know if this works for you
